src/plugins/profiling/server/routes/stacktrace.ts

```diff
@@ -23,36 +23,104 @@ import { getHitsItems, getDocs } from './compat';
-  frameIDToFileIDCache.set(fileIDChunk, fileID);
-  return fileID;
+const FILE_ID_LENGTH = 16;
+const FRAME_ID_LENGTH = 24;
+
+export interface EncodedStackTrace {
+  // This field is a base64-encoded byte string. The string represents a
+  // serialized list of frame IDs. Each frame ID is composed of two
+  // concatenated values: a 16-byte file ID and an 8-byte address or line
+  // number (depending on the context of the downstream reader).
+  //
+  //         Frame ID #1               Frame ID #2
+  // +----------------+--------+----------------+--------+----
+  // |     File ID    |  Addr  |     File ID    |  Addr  |
+  // +----------------+--------+----------------+--------+----
+  FrameID: string;
+
+  // This field is a run-length encoding of a list of uint8s. The order is
+  // reversed from the original input.
+  Type: string;
 }
 
-// extractFileIDArrayFromFrameIDArray extracts all FileIDs from the array of FrameIDs
-// and returns them as an array of base64url encoded strings. The order of this array
-// corresponds to the order of the input array.
-function extractFileIDArrayFromFrameIDArray(frameIDs: string[]): string[] {
-  const fileIDs = Array<string>(frameIDs.length);
-  for (let i = 0; i < frameIDs.length; i++) {
-    fileIDs[i] = extractFileIDFromFrameID(frameIDs[i]);
+// runLengthDecodeReverse decodes a run-length encoding for the reversed input array.
+//
+// The input is a binary stream of 2-byte pairs (first byte is the length and the
+// second byte is the binary representation of the object). The output is a list of
+// uint8s in reverse order.
+//
+// E.g. byte array [5, 0, 2, 2] is converted into an uint8 array like
+// [2, 2, 0, 0, 0, 0, 0].
+export function runLengthDecodeReverse(input: Buffer): number[] {
+  let size = 0;
+  for (let i = 0; i < input.length; i += 2) {
+    size += input[i];
   }
-  return fileIDs;
+
+  const output: number[] = new Array(size);
+
+  let i = input.length - 2;
+  let j = 0;
+  let count = input[i];
+
+  while (j < output.length) {
+    if (count > 0) {
+      output[j] = input[i + 1];
+      count--;
+      j++;
+    } else {
+      i -= 2;
+      count = input[i];
+    }
+  }
+  return output;
+}
+
+// decodeStackTrace unpacks an encoded stack trace from Elasticsearch
+export function decodeStackTrace(input: EncodedStackTrace): StackTrace {
+  const serializedFrameIDs = Buffer.from(input.FrameID, 'base64url');
+  const countFileAndFrameIDs = serializedFrameIDs.length / FRAME_ID_LENGTH;
+  const fileIDs: string[] = new Array(countFileAndFrameIDs);
+  const frameIDs: string[] = new Array(countFileAndFrameIDs);
+
+  // Step 1: Convert the serialized frameID list into two separate lists
+  // (frame IDs and file IDs). The first 16 bytes of a frame ID contains
+  // the FileID.
+  for (let i = 0; i < serializedFrameIDs.length; i += FRAME_ID_LENGTH) {
+    const frameIDBytes = serializedFrameIDs.slice(i, i + FRAME_ID_LENGTH);
+    const frameID = frameIDBytes.toString('base64url');
```

Comment:
This base64 encoding can be avoided when we split the serialized `input.FrameID` into chunks of 32 bytes for each frameID. Because the binary 24 bytes exactly translate into 32 bytes base64 :-)

Reply:
Nice, I changed the code so that decoding the original base64 wasn't needed.